### PR TITLE
lint: enforce some spelling conventions

### PR DESCRIPTION
**Pull request by bakkot**
This rule is tripped 29 times on ECMA-262 master: in 26 places where we say `*this* object` instead of (as in the other 429 uses of `*this*`) `*this*` value", in 1 place where we say `behavior` instead of (as in the other 132 places) `behaviour`, and in 2 places where we say `the empty string` instead of (as in the other 76 places) `the empty String`.

I'll fix those as part of upstreaming this.

### CODE REVIEW

**Comment by bakkot** [comment]
Note that it is plausible that we actually do want to say `*this* object`; @rkirsling [here](https://github.com/tc39/ecma262/pull/1736) observes that it appears to be a shorthand for "this value (which is an object)". It's only ever used to describe the behavior of built-in methods (e.g. [`String.prototype.concat`](https://tc39.es/ecma262/#sec-string.prototype.concat)). But I don't think it makes sense to have that distinction; in those cases nothing enforces that `*this*` is in fact an object.

File: src/lint/collect-spelling-diagnostics.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    pattern: /2's complement/giu,
+    message: 'Prefer "two\'s complement"',
+  },
+  {
+    pattern: /\*0\*/gu,
+    message: 'The Number value 0 should be written "*+0*", to unambiguously exclude "*-0*"',
+  },
+  {
+    pattern: /behavior/giu,
+    message: 'ECMA-262 uses UK spelling ("behaviour")',
```
Comment (rkirsling) [inline src/lint/collect-spelling-diagnostics.ts:27]:
```suggestion
    message: 'ECMA-262 uses Oxford spelling ("behaviour")',
```
nit: Probably worth being [precise](https://en.wikipedia.org/wiki/Oxford_spelling).

File: src/lint/collect-spelling-diagnostics.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import { offsetToLineAndColumn } from './utils';
+
+let ruleId = 'spelling';
+
+// Note that these will be composed, so cannot contain backreferences
+let matchers = [
+  {
+    pattern: /\*this\* object/giu,
+    message: 'Prefer "the *this* value"',
```
Comment (rkirsling) [inline src/lint/collect-spelling-diagnostics.ts:11]: I think the intent here (with `the` in the message but not in the pattern) is clear, so I don't object, but I will note that there are a few instances of `a *this* value` in the spec.

**Comment by ljharb** [comment]
I'd agree; the receiver only has to be an object in sloppy mode, and builtin methods have `RequireObjectCoercible` to account for nullish values, so `this value` seems to be the proper choice.

File: src/lint/collect-spelling-diagnostics.ts (hunk 2 above)
Comment (bakkot) [reply src/lint/collect-spelling-diagnostics.ts]: Removed the `the`.

File: src/lint/collect-spelling-diagnostics.ts (hunk 1 above)
Comment (bakkot) [reply src/lint/collect-spelling-diagnostics.ts]: Done, thanks.

File: src/lint/collect-spelling-diagnostics.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+          line,
+          column,
+          message,
+        });
+        match = pattern.exec(sourceText);
+      }
+    }
+    if (errors.length === 0) {
+      throw new Error(
+        'Ecmarkup has a bug: the spell checker reported an error, but could find one. Please report this at https://github.com/tc39/ecmarkup/issues/new.'
```
Comment (michaelficarra) [inline src/lint/collect-spelling-diagnostics.ts:58]: I don't understand this message. Is there a typo?

Comment (bakkot) [reply src/lint/collect-spelling-diagnostics.ts]: Ah, yeah, it should read "could *not*". Fixed.

File: src/lint/collect-spelling-diagnostics.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  {
+    pattern: /1's complement/giu,
+    message: 'Prefer "one\'s complement"',
+  },
+  {
+    pattern: /2's complement/giu,
+    message: 'Prefer "two\'s complement"',
+  },
+  {
+    pattern: /\*0\*/gu,
```
Comment (michaelficarra) [inline src/lint/collect-spelling-diagnostics.ts:22]:
```suggestion
    pattern: /\*0(?:<sub>𝔽</sub>)?\*/gu,
```

Comment (bakkot) [reply src/lint/collect-spelling-diagnostics.ts:22]: Eh, doesn't seem worth it.

**Pull request merged**